### ISSUE TITLE
Errors in upgrading to latest tornado (4.3.0)

### ISSUE THREAD

**Issue by delta24** [owtf/owtf#560]
Currently we have tornado==4.0.2 installed. Upgrading to the latest tornado may improve performance issues as well as add new capabilities. Upon installing 4.3.0 version, I got the following errors:

- [x] TypeError: <tornado.httputil.HTTPHeaders object at 0x7f3a53c7fd90> is not JSON serializable.

Below is the stacktrace.
```
[ERROR] [2016-02-29 11:59:28,315] [File 'web.py', line 1524, in log_exception] - Uncaught exception GET http://zero.webappsecurity.com/ (127.0.0.1)
HTTPServerRequest(protocol='http', host='zero.webappsecurity.com', method='GET', uri='http://zero.webappsecurity.com/', version='HTTP/1.1', remote_ip='127.0.0.1', headers={'Host': 'zero.webappsecurity.com', 'Proxy-Connection': 'Keep-Alive', 'Accept': '*/*', 'User-Agent': 'Mozilla/5.0 (X11; Linux i686; rv:6.0) Gecko/20100101 Firefox/15.0'})
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/tornado/web.py", line 1401, in _stack_context_handle_exception
    raise_exc_info((type, value, traceback))
  File "/usr/local/lib/python2.7/dist-packages/tornado/stack_context.py", line 314, in wrapped
    ret = fn(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/tornado/web.py", line 1613, in future_complete
    f.result()
  File "/usr/local/lib/python2.7/dist-packages/tornado/concurrent.py", line 232, in result
    raise_exc_info(self._exc_info)
  File "/usr/local/lib/python2.7/dist-packages/tornado/gen.py", line 1017, in run
    yielded = self.gen.send(value)
  File "/media/sf_owtf/framework/http/proxy/proxy.py", line 238, in get
    self.cache_handler.dump(response)
  File "/media/sf_owtf/framework/http/proxy/cache_handler.py", line 97, in dump
    json.dump(cache_dict, outfile)
  File "/usr/lib/python2.7/json/__init__.py", line 189, in dump
    for chunk in iterable:
  File "/usr/lib/python2.7/json/encoder.py", line 434, in _iterencode
    for chunk in _iterencode_dict(o, _current_indent_level):
  File "/usr/lib/python2.7/json/encoder.py", line 408, in _iterencode_dict
    for chunk in chunks:
  File "/usr/lib/python2.7/json/encoder.py", line 442, in _iterencode
    o = _default(o)
  File "/usr/lib/python2.7/json/encoder.py", line 184, in default
    raise TypeError(repr(o) + " is not JSON serializable")
TypeError: <tornado.httputil.HTTPHeaders object at 0x7f3a53c7fd90> is not JSON serializable
```

- [x] Filelock timeouts

This happens probably because of the above errors in the proxy.

Stacktrace:
```
[ERROR] [2016-02-29 12:05:19,127] [File 'web.py', line 1524, in log_exception] - Uncaught exception GET http://zero.webappsecurity.com/ (127.0.0.1)
HTTPServerRequest(protocol='http', host='zero.webappsecurity.com', method='GET', uri='http://zero.webappsecurity.com/', version='HTTP/1.1', remote_ip='127.0.0.1', headers={'Host': 'zero.webappsecurity.com', 'Proxy-Connection': 'Keep-Alive', 'Accept-Encoding': 'gzip, deflate', 'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8', 'User-Agent': 'Mozilla/5.0 (X11; Linux i686; rv:6.0) Gecko/20100101 Firefox/6.0'})
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/tornado/web.py", line 1401, in _stack_context_handle_exception
    raise_exc_info((type, value, traceback))
  File "/usr/local/lib/python2.7/dist-packages/tornado/stack_context.py", line 314, in wrapped
    ret = fn(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/tornado/web.py", line 1613, in future_complete
    f.result()
  File "/usr/local/lib/python2.7/dist-packages/tornado/concurrent.py", line 232, in result
    raise_exc_info(self._exc_info)
  File "/usr/local/lib/python2.7/dist-packages/tornado/gen.py", line 1017, in run
    yielded = self.gen.send(value)
  File "/media/sf_owtf/framework/http/proxy/proxy.py", line 120, in get
    self.cached_response = self.cache_handler.load()
  File "/media/sf_owtf/framework/http/proxy/cache_handler.py", line 117, in load
    self.file_lock.acquire()
  File "/media/sf_owtf/framework/lib/filelock.py", line 53, in acquire
    raise FileLockTimeoutException("%d seconds passed." % self.timeout)
FileLockTimeoutException: 30 seconds passed.
```

- [x] SSL errors

```
[ERROR] [2016-02-29 12:33:04,860] [File 'iostream.py', line 546, in _handle_events] - Uncaught exception, closing connection.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/tornado/iostream.py", line 507, in _handle_events
    self._handle_connect()
  File "/usr/local/lib/python2.7/dist-packages/tornado/iostream.py", line 1376, in _handle_connect
    do_handshake_on_connect=False)
  File "/usr/local/lib/python2.7/dist-packages/tornado/netutil.py", line 519, in ssl_wrap_socket
    return context.wrap_socket(socket, **kwargs)
  File "/usr/lib/python2.7/ssl.py", line 350, in wrap_socket
    _context=self)
  File "/usr/lib/python2.7/ssl.py", line 537, in __init__
    raise ValueError("check_hostname requires server_hostname")
ValueError: check_hostname requires server_hostname
[ERROR] [2016-02-29 12:33:04,861] [File 'ioloop.py', line 629, in handle_callback_exception] - Exception in callback None
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/tornado/ioloop.py", line 883, in start
    handler_func(fd_obj, events)
  File "/usr/local/lib/python2.7/dist-packages/tornado/stack_context.py", line 275, in null_wrapper
    return fn(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/tornado/iostream.py", line 507, in _handle_events
    self._handle_connect()
  File "/usr/local/lib/python2.7/dist-packages/tornado/iostream.py", line 1376, in _handle_connect
    do_handshake_on_connect=False)
  File "/usr/local/lib/python2.7/dist-packages/tornado/netutil.py", line 519, in ssl_wrap_socket
    return context.wrap_socket(socket, **kwargs)
  File "/usr/lib/python2.7/ssl.py", line 350, in wrap_socket
    _context=self)
  File "/usr/lib/python2.7/ssl.py", line 537, in __init__
    raise ValueError("check_hostname requires server_hostname")
ValueError: check_hostname requires server_hostname
```

- [x] StreamClosed errors. 

These exceptions should be properly handled in the code.
```
[ERROR] [2016-02-29 14:06:20,662] [File 'web.py', line 1524, in log_exception] - Uncaught exception CONNECT versioncheck-bg.addons.mozilla.org:443 (127.0.0.1)
HTTPServerRequest(protocol='http', host='versioncheck-bg.addons.mozilla.org:443', method='CONNECT', uri='versioncheck-bg.addons.mozilla.org:443', version='HTTP/1.1', remote_ip='127.0.0.1', headers={'Connection': 'keep-alive', 'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64; rv:38.0) Gecko/20100101 Firefox/38.0 Iceweasel/38.6.1', 'Host': 'versioncheck-bg.addons.mozilla.org:443', 'Proxy-Connection': 'keep-alive'})
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/tornado/web.py", line 1401, in _stack_context_handle_exception
    raise_exc_info((type, value, traceback))
  File "/usr/local/lib/python2.7/dist-packages/tornado/stack_context.py", line 314, in wrapped
    ret = fn(*args, **kwargs)
  File "/media/sf_owtf/framework/http/proxy/proxy.py", line 301, in ssl_fail
    self.request.connection.stream.write(b"HTTP/1.1 200 Connection established\r\n\r\n")
  File "/usr/local/lib/python2.7/dist-packages/tornado/iostream.py", line 377, in write
    self._check_closed()
  File "/usr/local/lib/python2.7/dist-packages/tornado/iostream.py", line 885, in _check_closed
    raise StreamClosedError(real_error=self.error)
StreamClosedError: Stream is closed
```

**Comment by delta24** [comment]
@tunnelshade can you help on the SSL errors? ;)

**Comment by delta24** [comment]
There is 1 more errors to be fixed:

- [ ] Error when loading the private key 
```
[E 160606 12:41:59 web:1524] Uncaught exception CONNECT www.google.co.in:443 (127.0.0.1)
    HTTPServerRequest(protocol='http', host='www.google.co.in:443', method='CONNECT', uri='www.google.co.in:443', version='HTTP/1.1', remote_ip='127.0.0.1', headers={'Host': 'www.google.co.in:443', 'Proxy-Connection': 'keep-alive', 'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.79 Safari/537.36'})
    Traceback (most recent call last):
      File "/usr/lib/python2.7/site-packages/tornado/web.py", line 1401, in _stack_context_handle_exception
        raise_exc_info((type, value, traceback))
      File "/usr/lib/python2.7/site-packages/tornado/stack_context.py", line 314, in wrapped
        ret = fn(*args, **kwargs)
      File "/home/viyat/workspace/owtf/framework/http/proxy/proxy.py", line 280, in start_tunnel
        success=ssl_success
      File "/home/viyat/workspace/owtf/framework/http/proxy/socket_wrapper.py", line 22, in wrap_socket
        key, cert = gen_signed_cert("*." + ".".join(domain.split(".")[-3:]), ca_crt, ca_key, ca_pass, certs_folder)
      File "/home/viyat/workspace/owtf/framework/http/proxy/gen_cert.py", line 44, in gen_signed_cert
        ca_key = crypto.load_privatekey(crypto.FILETYPE_PEM, open(ca_key).read(), passphrase=ca_pass)
      File "/usr/lib/python2.7/site-packages/OpenSSL/crypto.py", line 2483, in load_privatekey
        _raise_current_error()
      File "/usr/lib/python2.7/site-packages/OpenSSL/_util.py", line 48, in exception_from_error_queue
        raise exception_type(errors)
    Error: []
```

**Comment by delta24** [comment]
Closed in e1ba544f1fa9aaedd2d1c90b92ab3f27a6224905.
